### PR TITLE
[JENKINS-55070, 59992, 60454, 61192] Correct symptom of missing instance

### PR DESCRIPTION
**Pull request by Wadeck**
See [JENKINS-55070](https://issues.jenkins-ci.org/browse/JENKINS-55070), [JENKINS-59992](https://issues.jenkins-ci.org/browse/JENKINS-59992), [JENKINS-60454](https://issues.jenkins-ci.org/browse/JENKINS-60454) and [JENKINS-61192](https://issues.jenkins-ci.org/browse/JENKINS-61192).

For a detailed explanation of the (suspected?) root cause, see [this comment](https://issues.jenkins-ci.org/browse/JENKINS-59992?focusedCommentId=379193&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-379193).
**This PR only solves the reported symptom, not the root cause.**

⚠️ Any code called during a Jenkins restart cannot yet rely on `Jenkins.get()`.

### Proposed changelog entries

* Prevents one occurence of "Jenkins.instance is missing"

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [n/a] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

**Review by res0nance — APPROVED** [review]
LGTM for those reading the stack traces in the jira issues one of them was fixed in #3806

**Comment by Wadeck** [comment]
Thanks @res0nance for better knowing what I did recently than myself 🎉 I completely forgot that one

The two PRs are only covering symptoms :)

**Review by markjacksonfishing — APPROVED** [review]

**Comment by daniel-beck** [comment]
@Wadeck I do not understand how this addresses any of the referenced issues. Specifically, none of them appear to be timing-related, three of the four are fairly explicit that this is a pretty permanent situation once encountered, at least until the process is stopped. JENKINS-55070 (second comment) is very explicit about this happening after initiating a restart, and Jenkins being stopped. @oleg-nenashev's analysis seems to overlook the part where Jenkins has been stopped before the exception occurs.

I mean, if you just want to replace this particular exception with a different failure mode, that's fine. But the PR summary looks like you're actually resolving these issues, and I don't see how.

**Comment by Wadeck** [comment]
@daniel-beck I made this even more explicit in the description
from

> This PR only solves the symptom, not the root cause.

to

> **This PR only solves the reported symptom, not the root cause.**

**Comment by daniel-beck** [comment]
@Wadeck Thanks. I misunderstood "fix symptom" to be similar to https://github.com/jenkinsci/jenkins/pull/4170 which didn't actually fix the cause (https://github.com/jenkinsci/jenkins/pull/4456 did), but addressed the problems encountered because of it.

**Review by jeffret-b — APPROVED** [review]
My observation from before stands: since we're not fixing the root cause, it's hard to tell just how much this will help. However, this will probably improve some situations so it's worth a try. If we can narrow the problematic window sufficiently or collect specific enough information then hopefully we can eventually solve it.

File: core/src/main/java/hudson/security/TokenBasedRememberMeServices2.java [hunk 1]
```diff
@@ -161,10 +161,12 @@ public void loginSuccess(HttpServletRequest request, HttpServletResponse respons
 
     @Override
     public Authentication autoLogin(HttpServletRequest request, HttpServletResponse response) {
-        if(Jenkins.get().isDisableRememberMe()){
+        // as this filter could be called during restart, this corrects at least the symptoms
+        Jenkins j = Jenkins.getInstanceOrNull();
+        if (j != null && j.isDisableRememberMe()) {
```
Comment (daniel-beck) [inline core/src/main/java/hudson/security/TokenBasedRememberMeServices2.java]: It seems doubtful that without the `Jenkins` singleton, we have enough working that ending up in `retrieveAuthFromCookie` will ultimately be successful (`User#getById`?).

Just return `null` when `jenkins == null`, and return?

Comment (Wadeck) [reply core/src/main/java/hudson/security/TokenBasedRememberMeServices2.java]: Very good point, some implementation of the userDetailsService (like for the embedded SecurityRealm) are calling such mechanisms.

**Review by daniel-beck — APPROVED** [review]

**Comment by res0nance** [comment]
Will merge in 24h if no negative feedback

**Review by oleg-nenashev — APPROVED** [review]
Fine as a quick fix

**Comment by oleg-nenashev** [comment]
ATH is stuck due to the infra issue  I will ignore it and get the PR merged

**Pull request merged**